src/arm64/src/lj_str_hash_arm64.h

```diff
@@ -0,0 +1,271 @@
+  } else {
+	v1 = *cast_uint32p(str);
+	v2 = *cast_uint32p(str + len - 4);
+  }
+
+  h = __crc32cw(h, len);
+  h = __crc32cd(h, v1);
+  h = __crc32cd(h, v2);
+
+  return h;
+}
+
+static LJ_AINLINE uint32_t lj_str_hash_16_128(const char* str, size_t len)
+{
+  uint64_t h1 = 0, h2 = 0;
+  uint32_t i;
+
+  h1 = __crc32cw(h1, len);
+
+  for (i = 0; i < len - 16; i += 16) {
+	h1 += __crc32cd(h1, *cast_uint64p(str + i));
+	h2 += __crc32cd(h2, *cast_uint64p(str + i + 8));
+  }
+
+  h1 = __crc32cd(h1, *cast_uint64p(str + len - 16));
+  h2 = __crc32cd(h2, *cast_uint64p(str + len - 8));
+
+  return __crc32cw(h1, h2);
+}
+
+/* **************************************************************************
+ *
+ *  Following is code about hashing string with length >= 128
+ *
+ * **************************************************************************
+ */
+
+static uint32_t random_pos[32][2];
+static const int8_t log2_tab[128] = { -1,0,1,1,2,2,2,2,3,3,3,3,3,3,3,3,4,4,
+  4,4,4,4,4,4,4,4,4,4,4,4,4,4,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,
+  5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,6,6,6,6,6,6,6,6,6,6,6,6,
+  6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,
+  6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6 };
+
+/* return floor(log2(n)) */
+static LJ_AINLINE uint32_t log2_floor(uint32_t n)
+{
+  if (n <= 127) {
+    return log2_tab[n];
+  }
+
+  if ((n >> 8) <= 127) {
+    return log2_tab[n >> 8] + 8;
+  }
+
+  if ((n >> 16) <= 127) {
+    return log2_tab[n >> 16] + 16;
+  }
+
+  if ((n >> 24) <= 127) {
+    return log2_tab[n >> 24] + 24;
+  }
+
+  return 31;
+}
+
+#define POW2_MASK(n) ((1L << (n)) - 1)
+/* This function is to populate `random_pos` such that random_pos[i][*]
+ * contains random value in the range of [2**i, 2**(i+1)).
+ */
+static void arm64_init_random(void)
+{
+  int i, seed, rml;
+
+  /* Calculate the ceil(log2(RAND_MAX)) */
+  rml = log2_floor(RAND_MAX);
+  if (RAND_MAX & (RAND_MAX - 1)) {
+    rml += 1;
+  }
+
+  /* Init seed */
+  seed = 0;
+  seed = __crc32cw(seed, getpid());
+  seed = __crc32cw(seed, time(NULL));
+  srandom(seed);
+
+  /* Now start to populate the random_pos[][]. */
+  for (i = 0; i < 3; i++) {
+    /* No need to provide random value for chunk smaller than 8 bytes */
+    random_pos[i][0] = random_pos[i][1] = 0;
+  }
+
+  for (; i < rml; i++) {
+    random_pos[i][0] = random() & POW2_MASK(i+1);
+    random_pos[i][1] = random() & POW2_MASK(i+1);
+  }
+
+  for (; i < 31; i++) {
+    int j;
+    for (j = 0; j < 2; j++) {
+      uint32_t v, scale;
+      scale = random_pos[i - rml][0];
+      if (scale == 0) {
+         scale = 1;
+      }
+      v = (random() * scale) & POW2_MASK(i+1);
+      random_pos[i][j] = v;
+    }
+  }
+}
+#undef POW2_MASK
+
+void __attribute__((constructor)) arm64_init_constructor()
+{
+    // Check if crc32 supported.
+    unsigned long hwcap;
+    hwcap = getauxval(AT_HWCAP);
+    if (hwcap & HWCAP_CRC32) {
+	LJ_STR_HASH = lj_str_hash;
+    }
+    else {
+	LJ_STR_HASH = lj_str_original_hash;
```

Comment:
Suggest: Initialize LJ_STR_HASH to lj_str_hash and only if HWCAP_CRC32 is not set, reset it to lj_str_original_hash over here.